force_bdss/core_evaluation_driver.py

```diff
@@ -31,18 +31,96 @@ def application_started(self):
+                mco_data_values,
+                ds_model.input_slot_maps,
+                in_slots)
+
+            res = data_source.run(ds_model, passed_data_values)
+            if len(res) != len(out_slots):
+                raise RuntimeError("The number of data values returned by"
+                                   " the DataSource does not match the number"
+                                   " of parameters specified. This is likely a"
+                                   " DataSource plugin error.")
+
+            if len(res) != len(ds_model.output_slot_names):
+                raise RuntimeError("The number of data values returned by"
+                                   " the DataSource does not match the number"
+                                   " of names specified. This is either an"
+                                   " input file error or a plugin error.")
+
+            for dv, output_slot_name in zip(res, ds_model.output_slot_names):
+                dv.name = output_slot_name
+
+            ds_results.extend(res)
 
         kpi_results = []
         for kpic_model in workflow.kpi_calculators:
             kpic_bundle = kpic_model.bundle
             kpi_calculator = kpic_bundle.create_kpi_calculator()
-            kpi_results.extend(kpi_calculator.run(kpic_model, ds_results))
+            in_slots, out_slots = kpi_calculator.slots(kpic_model)
+
+            passed_data_values = self._bind_data_values(
+                mco_data_values+ds_results,
+                kpic_model.input_slot_maps,
+                in_slots)
+
+            res = kpi_calculator.run(kpic_model, passed_data_values)
+            if len(res) != len(out_slots):
+                raise RuntimeError("The number of data values returned by"
+                                   " the KPICalculator does not match the"
+                                   " number of parameters specified. This is"
+                                   " likely a KPICalculator plugin error.")
+
+            if len(res) != len(kpic_model.output_slot_names):
+                raise RuntimeError("The number of data values returned by"
+                                   " the KPICalculator does not match the"
+                                   " number of names specified. This is"
+                                   " either an input file error or a plugin"
+                                   " error.")
+
+            for kpi, output_slot_name in zip(res,
+                                             kpic_model.output_slot_names):
+                kpi.name = output_slot_name
+
+            kpi_results.extend(res)
 
         mco_communicator.send_to_mco(mco_model, kpi_results)
+
+    def _bind_data_values(self,
+                          available_data_values,
+                          model_slot_map,
+                          slots):
+
+        passed_data_values = []
+        lookup_map = {dv.name: dv for dv in available_data_values}
+
+        if len(slots) > len(model_slot_map):
```

Comment:
Even if in the case of `len(slots) < len(model_slot_map)` it won't result in a failure, we should consider it a FileError, no ?